Content.Shared/Clothing/EntitySystems/ClothingSystem.cs

```diff
@@ -85,49 +85,77 @@ private void QuickEquip(
         }
     }
 
-    private void ToggleVisualLayers(EntityUid equipee, HashSet<HumanoidVisualLayers> layers, HashSet<HumanoidVisualLayers> appearanceLayers)
+    private void ToggleVisualLayers(
+        EntityUid equipee,
+        EntityUid equipment,
+        HashSet<HumanoidVisualLayers> appearanceLayers
+        )
     {
-        foreach (HumanoidVisualLayers layer in layers)
+        if (TryComp(equipment, out ClothingComponent? clothing) &&
+            TryComp(equipment, out HideLayerClothingComponent? hideLayer))
         {
-            if (!appearanceLayers.Contains(layer))
-                continue;
-
-            InventorySystem.InventorySlotEnumerator enumerator = _invSystem.GetSlotEnumerator(equipee);
+            if (hideLayer.Layers.Count > 0)
+            {
+                // iterate the HideLayerClothingComponent's layers map and check that
+                // the clothing is equipped in a matching slot.
+                foreach (HumanoidVisualLayers layer in hideLayer.Layers.Keys)
+                {
+                    if (!appearanceLayers.Contains(layer))
+                        continue;
 
-            bool shouldLayerShow = true;
-            while (enumerator.NextItem(out EntityUid item, out SlotDefinition? slot))
+                    bool shouldLayerShow = true;
+                    SlotFlags hideSlot = hideLayer.Layers[layer];
+                    if (clothing.EquippedInSlot == hideSlot)
+                    {
```

Comment:
IMO, this should support more than one slot. I.e., it should optionally allow you to hide the same layer even when equipped to a number of different slots.

```suggestion
                    var validSlots = hideLayer.Layers[layer];
                    if (validSlots.HasFlag(clothing.EquippedInSlot))
                    {
```

Reply:
Actually there's another issue here. `ClothingComponent.EquippedInSlot` gets cleared by `OnGotUnequipped` before `CheckEquipmentForLayerHide` is raised.

The overall logic would also just be cleaner if layer hiding were handled by a separate system that listens for `ClothingGotUnequippedEvent`, `ItemMaskToggledEvent` and `ClothingGotUnequippedEvent` events directed at entities that have a HideLayerClothingComponent.